mii/grpc_related/modelresponse_server.py

```diff
@@ -44,120 +44,48 @@ def _get_model_time(self, model, sum_times=False):
             model_time = -1
         return model_time
 
-    def GeneratorReply(self, request, context):
-        query_kwargs = self._unpack_proto_query_kwargs(request.query_kwargs)
+    def _run_inference(self, method_name, request_proto):
+        if method_name not in self.method_name_to_task:
+            raise ValueError(f"unknown method: {method_name}")
+
+        task = self.method_name_to_task[method_name]
+        if task not in GRPC_METHOD_TABLE:
+            raise ValueError(f"unknown task: {task}")
 
-        # unpack grpc list into py-list
-        request = [r for r in request.request]
+        conv_funcs = GRPC_METHOD_TABLE[task]
+        args, kwargs = conv_funcs["unpack_request_from_proto"](request_proto)
 
         start = time.time()
-        batched_responses = self.inference_pipeline(request, **query_kwargs)
+        response = self.inference_pipeline(*args, **kwargs)
         end = time.time()
 
-        # response is a list
-        text_responses = []
-        for response in batched_responses:
-            text = response[0]['generated_text']
-            text_responses.append(text)
+        model_time = self._get_model_time(self.inference_pipeline.model,
+                                          sum_times=True) if hasattr(
+                                              self.inference_pipeline,
+                                              "model") else -1
 
-        model_time = self._get_model_time(self.inference_pipeline.model, sum_times=True)
+        return conv_funcs["pack_response_to_proto"](response, end - start, model_time)
 
-        val = modelresponse_pb2.MultiStringReply(response=text_responses,
-                                                 time_taken=end - start,
-                                                 model_time_taken=model_time)
-        return val
+    def GeneratorReply(self, request, context):
+        return self._run_inference("GeneratorReply", request)
```

Comment:
This refactor has me wondering if these functions are really necessary. For example, what if we just passed the `TASK_NAME` value to `_run_inference` from here? https://github.com/microsoft/DeepSpeed-MII/blob/b86e0ce9dbffe1ca75f7f9bd44a5f96a3999542e/mii/server_client.py#L267

Additionally, I think the way we currently pass `request_dict` and `query_kwargs` to [query](https://github.com/microsoft/DeepSpeed-MII/blob/b86e0ce9dbffe1ca75f7f9bd44a5f96a3999542e/mii/server_client.py#L349) could be simplified to `*args, **kwargs`. These args and kwargs would then be passed all the way to the pipeline object. This way users would run inference on MII with the same args and kwargs they would use with a `transformers.pipeline` object.

Reply:
I understand your point and agree that the current change looks redundant.

My concern is about serialization. If you use the same gPRC API for all tasks, the message needs to be able to contain any type of request and such a message is not very efficient. It seems natural that we have different message schemas for texts and images. This is also related to type hints. Once you use the same API for all tasks, you cannot give a specific data type to the request/response to a task type.

To reduce the redundancy of the method names, another approach could be to set a hook on the superclass, which is generated from the proto, to call `_run_inference` based on the method name. But I'm also wondering if it might be too complicated to use some reflection-like approach for this refactoring work.

I would be happy if you could give me any feedback.

Reply:
We will offer APIs that accept the same inputs as those to `transformers.pipeline` in the future, but not include the change in this PR.